File: src/main/scala/repository/eventsourcing/EventSourcedRepository.scala
package repository.eventsourcing

import java.util.Optional
import java.lang.Math.min
import java.lang.reflect.{Constructor, ParameterizedType}
import java.util.ConcurrentModificationException

import com.typesafe.scalalogging.LazyLogging
import eventstore.api.{Event, EventStore}
import repository.IdentifiedEntity
import scala.collection.JavaConverters._
import scala.compat.java8.OptionConverters._
import scala.compat.java8.StreamConverters._

abstract class EventSourcedRepository[T <: EventSourcedEntity[T] with IdentifiedEntity[K], K](val eventStore: EventStore)
  extends TemporalPersistenceOrientedRepository[T, K] with LazyLogging {
  override def get(id: K): Optional[T] = {
    get(id, -1)
  }

  private def get(id: K, version: Long): Optional[T] = {
    getByStreamName(streamName(id), version, snapshot(id, version)).asJava
  }

  private def getByStreamName(streamName: String, version: Long, snapshot: Option[T] = None): Option[T] = {
    eventStore.streamSince(streamName, snapshot.map(e => e.getUnmutatedVersion()).getOrElse(-1)).asScala.map(_.toScala[Stream]).flatMap { events =>
      if (events.isEmpty || events.lastOption.exists(_.isInstanceOf[RemovedEvent[K]])) {
        return snapshot
      }
      var entity = snapshot.getOrElse(init(events.head))
      var eventsRemaining = if (snapshot.isEmpty) events.tail else events
      while (entity.getMutatedVersion() != version && eventsRemaining.nonEmpty) {
        entity = entity.apply(eventsRemaining.head)
        eventsRemaining = eventsRemaining.tail
      }
      Some(entity.commitChanges())
    }
  }

  protected def saveSnapshot(entity: T): Unit = {}

  protected def removeSnapshot(id: K): Boolean = { false }

  protected def snapshot(id: K, before: Long): Option[T] = { None }

  private def init(initEvent: Event): T = {
    initEvent.asInstanceOf[InitialEvent[T]].initializedObject()
  }

  private def getAndApply(id: K, after: Long, changes: List[Event]): Option[T] = {
    if (after <= 0) {
      getAndApply(id, 1, changes.tail)
    } else {
      get(id, after).asScala match {
        case Some(entity) =>
          var mutatedEntity = entity.asInstanceOf[T]
          eventStore.streamSince(streamName(id), after).asScala.map(_.toScala[Stream]).flatMap { events =>
            var i: Int = 0
            while (min(changes.length, events.length) > i && changes(i).equals(events(i))) {
              mutatedEntity = mutatedEntity.apply(events(i))
              i += 1
            }
            events.takeRight(events.length - i).foreach { e => mutatedEntity = mutatedEntity.apply(e) }
            mutatedEntity = mutatedEntity.commitChanges()
            changes.takeRight(changes.length - i).foreach { e => mutatedEntity = mutatedEntity.apply(e) }
            Some(mutatedEntity)
          }
        case None => None
      }
    }
  }

  override def save(entity: T): T = {
    if (!entity.getChanges().iterator().hasNext()) {
      entity
    } else {
      try {
        eventStore.append(streamName(entity.getId()), entity.getUnmutatedVersion(), entity.getChanges())
        try {
          saveSnapshot(entity)
        } catch {
          case e: Exception => logger.error("Couldn't save snapshot", e)
        }
      } catch {
        case e: ConcurrentModificationException =>
          val freshEntity = getAndApply(entity.getId(), entity.getUnmutatedVersion(), entity.getChanges().asScala.toList).get
          if (freshEntity.getUnmutatedVersion() == entity.getUnmutatedVersion()) {
            throw new IllegalStateException(s"Couldn't resolve the conflict, the saved entity ${entity} (${entity.getUnmutatedVersion()}, ${entity.getChanges()}), but got fresh ${freshEntity} (${eventStore.version(streamName(entity.getId()))}).", e)
          }
          save(freshEntity)
      }
      entity
    }
  }

  override def size: Long = {
    eventStore.size
  }

  override def all : java.util.Set[T] = {
    eventStore.streamNames.asScala.flatMap(s => getByStreamName(s, -1)).toSet.asJava
  }

  protected def entityClass: Class[T] = {
    this.getClass
      .getGenericSuperclass
      .asInstanceOf[ParameterizedType]
      .getActualTypeArguments()(0)
      .asInstanceOf[Class[T]]
  }

  override def remove(id: K): Boolean = {
    if (!contains(id)) {
      return false
    }
    try {
      eventStore.append(streamName(id), new RemovedEvent[K](id))
      removeSnapshot(id)
      true
    } catch {
      case e: ConcurrentModificationException => remove(id)
    }
  }

  /**
   * Whether the repository had the given entity.
   * @param id id of the entity.
   * @return true, if it contained the entity.
   */
  override def contained(id: K): Boolean = eventStore.contains(streamName(id))

  private def constructor(eventClass: Class[_ <: Event]): Constructor[T] = {
    val c =  entityClass
      .getConstructor(eventClass)
      .asInstanceOf[Constructor[T]]
    c.setAccessible(true)
    c
  }

  protected def streamName(id: K): String = {
     streamPrefix + id
  }

  private def streamPrefix: String = {
    this.entityClass.getSimpleName
  }
}

File: src/main/java/repository/eventsourcing/EventSourcedRepository.java
package repository.eventsourcing;

import java.util.Iterator;
import java.util.stream.Stream;
import java.util.Set;
import java.util.Optional;
import java.lang.reflect.Constructor;
import java.lang.reflect.ParameterizedType;
import java.util.ConcurrentModificationException;

import eventstore.api.Event;
import eventstore.api.EventStore;
import repository.IdentifiedEntity;

import static java.util.stream.Collectors.toSet;
import static java.util.stream.Collectors.toList;

public abstract class EventSourcedRepository<T extends EventSourcedEntity<T> & IdentifiedEntity<K>, K>
    implements TemporalPersistenceOrientedRepository<T, K> {

    protected EventStore eventStore;
    public final Class<T> entityClass;

    public EventSourcedRepository(EventStore eventStore) {
        this.eventStore = eventStore;
        entityClass = (Class<T>) ((ParameterizedType) this.getClass()
                .getGenericSuperclass())
                .getActualTypeArguments()[0];
    }

    @Override
    public Optional<T> get(K id) {
        return get(id, -1);
    }

    private Optional<T> get(K id, long version)  {
        return getByStreamName(streamName(id), version, snapshot(id, version));
    }

    private Optional<T> getByStreamName(String streamName, long version) {
        return getByStreamName(streamName, version, Optional.empty());
    }

    private Optional<T> getByStreamName(String streamName, long version, Optional<T> snapshot) {
        return eventStore.streamSince(streamName, snapshot.map((e) -> e.getUnmutatedVersion()).orElse(-1L))
            .map(e -> e.iterator())
            .flatMap((events) -> {
            if (!events.hasNext()) {
                return snapshot;
            }
            T entity = (T) (snapshot.isPresent() ? snapshot.get() : init((InitialEvent) events.next()));
            while (entity.getMutatedVersion() != version && events.hasNext()) {
                Event event = events.next();
                if (event instanceof RemovedEvent) {
                    return Optional.empty();
                }
                entity = entity.apply(event);
            }
            return Optional.of(entity.commitChanges());
        });
    }

    protected void saveSnapshot(T entity) {}

    protected boolean removeSnapshot(K id) { return false; }

    protected Optional<T> snapshot(K id, long before) { return Optional.empty(); }

    private T init(Event initEvent) {
        return ((InitialEvent<T>) initEvent).initializedObject();
    }

    private Optional<T> getAndApply(K id, long after, Stream<Event> changes) {
        if (after <= 0) {
            return getAndApply(id, 1, changes.skip(1));
        } else {
            return get(id, after).flatMap((entity) -> {
                return eventStore.streamSince(streamName(id), after).map((e) -> e.iterator()).map((events) -> {
                    T mutatedEntity = entity;
                    int i = 0;
                    Iterator<Event> changesIterator = changes.iterator();
                    Event newEvent = null;
                    Event savedEvent = null;
                    while(events.hasNext()) {
                        mutatedEntity = mutatedEntity.apply(savedEvent = events.next());
                        // Trying to find the event, where the stream starts to differ from the saved one.
                        if (newEvent == null) {
                            newEvent = changesIterator.next();
                            if (newEvent.equals(savedEvent)) {
                                newEvent = null;
                            }
                        }
                    }
                    mutatedEntity = mutatedEntity.commitChanges();
                    // Do not forget to apply the first different event we found,
                    // since it's already read from the iterator.
                    if (newEvent != null) {
                        mutatedEntity = mutatedEntity.apply(newEvent);
                    }
                    while(changesIterator.hasNext()) { mutatedEntity = mutatedEntity.apply(changesIterator.next()); }
                    return mutatedEntity;
                });
            });
        }
    }

    @Override
    public T save(T entity) {
        if (entity.getChanges().isEmpty()){
            return entity;
        } else {
            try {
                eventStore.append(streamName(entity.getId()), entity.getUnmutatedVersion(), entity.getChanges());
                saveSnapshot(entity);
            } catch (ConcurrentModificationException e) {
                try {
                    T freshEntity = getAndApply(entity.getId(), entity.getUnmutatedVersion(), entity.getChanges().stream()).get();
                    if (freshEntity.getUnmutatedVersion() == entity.getUnmutatedVersion()) {
                        throw new IllegalStateException(
                                String.format(
                                    "Couldn't resolve the conflict, the saved entity %s (%s, %s), but got fresh %s (%s).",
                                    entity,
                                    entity.getUnmutatedVersion(),
                                    entity.getChanges(),
                                    freshEntity,
                                    eventStore.version(streamName(entity.getId()))), e);
                    }
                    return save(freshEntity);
                } catch (EventSourcingException esException) {
                    throw new EventSourcingConflictException("Couldn't resolve the conflict", esException);
                }
            }
            return entity;
        }
    }

    @Override
    public long size() {
        return eventStore.size();
    }

    @Override
    public Set<T> all() {
        return eventStore
            .streamNames()
            .stream()
            .map(s -> getByStreamName(s, -1L))
            .flatMap(s -> s.map(Stream::of).orElse(Stream.empty()))
            .collect(toSet());
    }

    @Override
    public boolean remove(K id) {
        if (!contains(id)) {
            return false;
        }
        try {
            eventStore.append(streamName(id), new RemovedEvent<K>(id));
            removeSnapshot(id);
            return true;
        } catch(ConcurrentModificationException e) {
            return remove(id);
        }
    }

    /**
     * Whether the repository had the given entity.
     * @param id id of the entity.
     * @return true, if it contained the entity.
     */
    @Override
    public boolean contained(K id) { return eventStore.contains(streamName(id)); }

    protected String streamName(K id) {
        return this.entityClass.getSimpleName() + id;
    }
}
